core/src/main/java/com/graphhopper/routing/Router.java

```diff
@@ -207,7 +207,9 @@ protected GHResponse routeRoundTrip(GHRequest request, FlexSolver solver) {
 
         RoundTripRouting.Result result = RoundTripRouting.calcPaths(snaps, pathCalculator);
         // we merge the different legs of the roundtrip into one response path
-        ResponsePath responsePath = concatenatePaths(request, solver.weighting, queryGraph, result.paths, getWaypoints(snaps));
+        // note that the waypoints are not just the snapped points of the snaps, as usual, because we do some kind of tweak
+        // to avoid 'unnecessary tails' in the roundtrip algo
+        ResponsePath responsePath = concatenatePaths(request, solver.weighting, queryGraph, result.paths, result.wayPoints);
```

Comment:
I don't really understand why we do this tweak, but if we do we should adjust the `snapped_waypoints` as well. The `snapped_waypoints` are quite meaningless for round trip, though, as the via-points are generated by the algorithm.